File: perl/xmms2.pl
# A script which print currently played track (format is configurable) in xmms2 to a channel or private window.
# This script is BSD licensed.
# Konstantin Merenkov @ 17.09.2007 11:20:57 MSD
# kmerenkov AT gmail DOT com
#
# changelog:
# 30.09.2007 23:53:12 MSD: 0.2: added xmms2print, which doesn't send information to a channel but just prints it to a buffer
# 17.09.2007 11:20:57 MSD: 0.1: initial version, can send information about the played song to a channel

use strict;
use warnings;
use Audio::XMMSClient;

weechat::register("xmms2", "0.2", "", "Prints currently played track info to a channel");

weechat::add_command_handler("xmms2", "xmms2", "Sends currently played track info to a channel");
weechat::add_command_handler("xmms2print", "xmms2print", "Prints currently played track to a buffer");

sub get_npformat {
    my $np_format = weechat::get_plugin_config("npformat");
    if (!$np_format) {
        # format is pretty simple:
        # just prefix any key returned from medialib_get_info with $ to get it replaced.
        # Example: np: $artist - $album ($date) - $title
        $np_format = 'np: $artist - $title';
        if (weechat::set_plugin_config("npformat", $np_format) == 0) {
            weechat::print("Error on setting xmms2.npformat option");
        }
    }
    return $np_format;
}

sub xmms2 {
    my $retval = get_line();
    if (length($retval) > 0) {
        weechat::command("/me $retval");
        return weechat::PLUGIN_RC_OK;
    }
    return weechat::PLUGIN_RC_KO;
}

sub xmms2print {
    my $retval = get_line();
    weechat::print("xmms2: $retval");
    return weechat::PLUGIN_RC_OK;
}

sub get_line {
    my $xmms = Audio::XMMSClient->new("weechat-script");
    if (!$xmms->connect) {
        weechat::print("Connection to xmms2 failed: ".$xmms->get_last_error);
    }
    else {
        my $result = $xmms->playback_current_id;
        $result->wait;
        if ($result->iserror) {
            weechat::print("Playback current id returned error: ".$result->get_error);
        }
        else {
            my $id = $result->value;
            if ($id == 0) {
                weechat::print("Nothing is playing in xmms2");
            }
            else {
                $result = $xmms->medialib_get_info($id);
                $result->wait;
                if ($result->iserror) {
                    weechat::print("Error while getting information about currently playing song");
                }
                else {
                    # here's a small caveat
                    # if you don't have 'date' key returned (i.e. it is not in metadata for your currently playing song
                    # and have $date in format string, it won't be replaced (keep your metadata in a good state :p)
                    # it applies to any keyword, not only 'date'
                    my $info = $result->value;
                    my $format = get_npformat;
                    foreach my $key (keys %{$info}) {
                        $format =~ s/\$$key/$info->{$key}/g;
                    }
                    return $format;
                }
            }
        }
    }
    return "";
}

File: ruby/xmms2.rb
# -*- coding: utf-8 -*-
#
# Copyright (c) 2009, 2011 Łukasz P. Michalik <lmi@ift.uni.wroc.pl>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2, as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
# USA.
#

#
# This plugin requires Xmms2 ruby bindings to be installed.  'format'
# configuration variable specifies how metadata is displayed.
# Default_format is a good example, but don't abuse the syntax as you
# might get garbage.  It works with any key found in medialib for a
# current ID (see `xmms2 info` for a list); non-existing keys are
# ignored.
#

require 'xmmsclient'

Default_format = "[np] ${title} - ${artist}"

$xmms = nil

def weechat_init
    Weechat.register('xmms2', 'Łukasz P. Michalik <lmi@ift.uni.wroc.pl>', '0.2', 'GPL2', 'Xmms2 interaction plugin', '', '')
    Weechat.hook_command('xmms2', 'Trigger a /me currently playing info', '', '', '', 'xmms2', '')
    if (Weechat.config_get_plugin("format").empty?)
        Weechat.config_set_plugin("format", Default_format)
    end
    return Weechat::WEECHAT_RC_OK
end

def xmms2(data, buffer, args)
    if not $xmms
        $xmms = Xmms::Client.new('weechat')

        begin
            $xmms.connect(ENV['XMMS_PATH'])
        rescue Xmms::Client::ClientError
            Weechat.print("", "Failed to connect to XMMS2 daemon.")
            return Weechat::WEECHAT_RC_ERROR
        end
    end

    begin
        id = $xmms.playback_current_id.wait.value

        if (id == 0)
            Weechat.print(buffer, "Nothing is played!")
            return Weechat::WEECHAT_RC_OK
        end
    rescue Xmms::Rescue::ValueError
        Weechat.print(buffer, "Error retrieving current ID!")
        return Weechat::WEECHAT_RC_ERROR
    end

    begin
        info = $xmms.medialib_get_info(id).wait.value.to_propdict
        reg = /\$\{(\w+)\}/
        format = Weechat.config_get_plugin("format")
        out = ""
        while m = reg.match(format)
            out << m.pre_match
            format = m.post_match
            key = m[1].intern
            if info.has_key? key
                out << info[key]
            end
        end
    rescue Xmms::Result::ValueError
        puts 'There was an error retrieving mediainfo for the current ID.'
    end

    Weechat.command(buffer, "/me #{out}")
    return Weechat::WEECHAT_RC_OK
end
